### ISSUE TITLE
load_setup_py_data() support

### ISSUE THREAD

**Issue by mparry** [mamba-org/boa#277]
From some evaluation of `boa`, my thoughts are a) it's wildly faster than `conda-build`, and b) we should be able to, and would love to, switch to it for most/all of our packages. Unfortunately, as it stands, that would currently be blocked in many cases by the lack of support for `load_setup_py_data()` ([docs](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#templating-with-jinja)) or equivalent.

Is this something that you would be open to supporting in some form? I think it would also require some looping and maybe conditional support along the lines of what you outline in https://github.com/mamba-org/boa/issues/191#issuecomment-935540102

I think that the support for providing the data could be lifted directly from `conda-build` [here](https://github.com/conda/conda-build/blob/master/conda_build/_load_setup_py_data.py). I'd be happy to provide a PR for that, but adding other templating constructs would presumably be a bit more involved.

**Comment by wolfv** [comment]
Hi @mparry thanks for the kind words and filing this issue. 

If I understand the functionality correctly this feature would need to download the source, then extract data from setup.py and then continue with the further evaluation of the YAML document, right?

I see two options:

- If we encounter this function in the `context` section, we unconditionally download all sources, with the expectation that for the source section we don't need to have any other variables than those set _earlier_ (e.g. version / name interpolations in the source URLs. We could download the source with the usual mechanism and extract it into a temporary folder that we remove again after finishing the extraction from setup.py. The source .tar.gz etc. files are cached independently so for small packages this wouldn't be a big overhead and thanks to the cache we don't have to re-download the files.
- You could write some script that generates the `recipe.yaml` files based on the `setup.py` file. Generating these files is more straightforward as you may know since it's just YAML. Not sure if this is an option you'd consider.

**Comment by mparry** [comment]
Yes, good point - it is a bit more complicated than I was originally suggesting, in that I agree that some multi-stage process results.

We certainly could deal with this ourselves, externally to `boa`, probably based upon that same `conda-build` function. I think it would have some downsides though - e.g. building a package locally outside CI would be more awkward.

As I say, I would like to offer some help with this but implementing the (sensible-sounding) solution you outline in your first bullet point might involve deeper changes than would be very practical for me to take on.

I don't know how widely used this `conda-build` feature is, and therefore of how much general interest it might be for `boa`. I suppose one way of evaluating that would be searching here in GitHub, for which there seems to be ~1.5k hits in recipe files: https://github.com/search?l=YAML&q=load_setup_py_data&type=Code

**Comment by wolfv** [comment]
I think the first solution won't take that much effort. 

This is the source download function: https://github.com/mamba-org/boa/blob/fe4d93b2ad7b323d46e770962e9aa5848a6379c4/boa/core/build.py#L554

**Comment by dhirschfeld** [comment]
Is there a reason to use `load_setup_py_data` other than setting the version? If it's just for that, there may be better solutions.

As a general-purpose solution, `load_setup_py_data` has a critical flaw (at least as-implemented) in that missing `setup.py` dependencies can prevent the recipe from even being parsed:
https://github.com/conda/conda-build/issues/2458#issuecomment-424906994

**Comment by mparry** [comment]
Well, I can't speak for the 1.5k usages in GitHub! However, our main use case is to share setup/config between the conda/boa recipe and things like tox.

**Comment by wolfv** [comment]
@dhirschfeld right, that's exactly the kind of "recursive" chicken - and - egg situation that I would love to avoid.

Ideally instead of reading a setup.py, one would specify dependencies in a `toml` file or similar static metadata file and generate the `boa` and `pypi` "recipes" from that.

Or, you coudl use the `setup.py` file to simply parse the dependencies out of the `boa` recipe (since it's _pure_ yaml it's straightforward!)

**Comment by dhirschfeld** [comment]
> Ideally instead of reading a setup.py, one would specify dependencies in a `toml` file or similar static metadata file and generate the `boa` and `pypi` "recipes" from that.

It seems this functionality is now part of `conda-build` as `load_file_data`:
* https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#loading-data-from-other-files

Added in https://github.com/conda/conda-build/pull/4480

@wolfv - is it planned to support `load_file_data`? If not, would that be a reasonable feature request?

**Comment by dhirschfeld** [comment]
Perhaps made redundant by https://github.com/mamba-org/boa/pull/345?
